1-Draft/RFCNNNN-Side-By-Side-Support.md

```diff
@@ -0,0 +1,253 @@
+6.2.1, with an error indicating that 6.2.0 was not installed).
+
+The UI installers should provide options to upgrade the latest installed
+version of PowerShell or install a new version side-by-side.
+
+### Listing installed versions of PowerShell
+
+```bash
+# This launches the latest version of PowerShell, returns a list of installed
+# versions and their installation location, and then exits
+pwsh -ListVersions
+# This does the same thing
+pwsh -lv
+```
+
+```output
+6.2.1 [/usr/local/microsoft/powershell/6.2.1/]
+7.0.100-preview1 [/usr/local/microsoft/powershell/7.0.100-preview1]
+```
+
+```powershell
+
+# This returns a list of installed versions and their installation location as
+# objects
+Get-InstalledPSVersion
+```
+
+```output
+Version                    InstallLocation
+-------                    ---------------
+6.2.1                      C:\ProgramFiles\PowerShell\6.2.1
+7.0.100-preview1           C:\ProgramFiles\PowerShell\7.0.100-preview1
+```
+
+### Launching a specific version of PowerShell
+
+```bash
+# This launches the latest version of PowerShell, gets the installation
+# location of the specified version, launches pwsh for that version, and exits
+# when the launched version exits
+pwsh -Version 6.2.1
+# This does the same thing
+pwsh -v 6.2.1
+```
+
+```output
+# A new window with the desired version of PowerShell open.
+# If it is possible to do have this window on a new tab in the new Terminal,
+# that would be great.
+```
+
+## Specification
+
+* create a `Get-InstalledPSVersion [<CommonParameters>]` cmdlet that returns
+the installed versions of PowerShell along with their locations
+* update the `-version` argument of `pwsh.exe` to attempt to launch a specific
+version of PowerShell
+* add a `-ListVersions` argument to `pwsh.exe` that internally invokes
+`Get-InstalledPSVersion`, formats and outputs the results, and exits
+* update the release packaging process to produce versioned packages
+
+## Alternate Proposals and Considerations
+
+### Include Windows PowerShell version support
+
+It would be worth considering having `Get-InstalledPSVersion` return the
+installed version of Windows PowerShell along with its location, and having
+`pwsh -version` support launching Windows PowerShell. That would allow users to
+launch any installed PowerShell version from `pwsh.exe`, making it easy to set
+up automation with one command to invoke. Arguments from `pwsh.exe` could be
+mapped to arguments on `powershell.exe`.
+
+### Are containers and Docker enough to solve this need?
+
+While discussing this with a colleague of mine, they asked if this is really
+needed, or if containers and Docker solve this problem already. While I believe
+containers can allow users to use different versions of PowerShell, I also
+suspect that most users don't know how to set that up, if they even use Docker
+at all. I think it would be easier for everyone if we could just launch the
+proper version of PowerShell via `pwsh.exe`.
```

Comment:
Who is your audience?

I'm an admin scripter. And while I know what Docker/containers are, I certainly don't use them, nor do I know how (I have done a Windows container lab, but not using Docker). Lots of desktops/laptops don't have virtualization turned on, nor have a virtualizer installed nor will your average user know how to do so. But their admins may have very good reasons to need multiple versions of PS.